1. 3 will be printed because i was declared as a var in line 6, so i is accessable outside of the scope. Then i is incremented 3 times because the length of the prices array that is passed in as the parameter is 3, so i = 3 after the loop.
2. 150 will be printed by line 13 the discountedPrice stores the last calculation result in the loop because it is updated every loop. Hence it would be 300 * (1-0.5) which is 150. The discountedPrice is also accessable outside of the scope because it is declared by var.
3. 150 will be printed by line 14 because finalPrice basically will store the same value as discountedPrice if the discountedPrice is a integer. Then finalPrice was declared as a var in the beginning at line 4, hence it is accessible outside of the scope, so it stores the same value as discountedPrice and will print out 150.
4. The function will return an array [50, 100, 150] because essentiall the function will take in the price array and multiply it by the discount parameter. In this case it is [100, 200, 300] and discount is 0.5. Hence it will multiply each of the integers in the array by 0.5, and push it into the discounted array, which results in [50, 100, 150], and then this array will be returned by the function.
5. It will return an error because i is defined by let in the loop, hence it is not accessible outside of its scope and we can't access it outside of the for loop, and it will return an error when we try to print it.
6. Same reason as number 5, because discountedPrice is declared with let, we are not able to access it outside of the scope, so that means it is only accessable inside the for loop but not outside.
7. 150 will be printed out by line 14 because although it was declared with let, it was declared outside of the for loop and inside the same scope as line 14. That means line 14 can access the finalPrice variable. Then it goes through the same process as question 3 and stores the most recent calculated finalPrice value inside the for loop, which is 150 in this case.
8. The function will return the array [50, 100, 150] with the same reason as question 4. Although this time variables are declared by let instead of var, but all the variables are used inside the block scope that it is in, hence it will not cause an error and just perform the same calculation as the one in question 4 and the push them into the discounted array and gets returned.
9. It will run in to an error because in line 6, we declared i by let, that means it will not be accessable outside its scope, which after the for loop, we are not able to access it anymore, so it will cause an error.
10. It will print out 3 because length is declared as a const variable with the length of the input prices array, in this case it is 3. After the declaration, we did not make any changes to the length variable, hence it will remain as 3. Although it is declared as a const varaible, it is declared inside the same scope as the console.log() function, so we will not have an error trying to print the length variable out.
11. It will return the array [50, 100, 150], although the discounted array is declared by const, it is in the same scope as all the other parts of the function, hence there will be no problem pushing element into this array. The function itself follows the same process as question 4 and 8, it does a calculation based on the input ([100, 200, 300], 0.5), multiplying each of the element in the input array by the 1 - discount input. Hence at the end, discounted will be [50, 100, 150] and then get returned by the function.
12. a. student.name
    b. student['Grad Year']
    c. student.greeting()
    d. student['Favorite Teacer'].name
    student.courseLoad[0]